### PR TITLE
FLPATH-377: workflow log

### PR DESCRIPTION
**Pull request by RichardW98**
**What this PR does / why we need it**:
display custom log in ui
**Which issue(s) this PR fixes** 
Fixes #[FLPATH-377](https://issues.redhat.com//browse/FLPATH-377)

**Change type**
- [ ] New feature
- [ ] Bug fix
- [ ] Unit tests
- [ ] Integration tests
- [ ] CI
- [ ] Documentation
- [ ] Auto-generated SDK code

**Impacted services**
- [ ] Workflow Service
- [ ] Notification Service

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.

### CODE REVIEW

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.redhat.parodos.workflow.enums;
+
+import lombok.Getter;
+
+public enum LogType {
+
+	TEXT(""), INFO("\u001B[32m"), WARNING("\u001B[33m"), ERROR("\u001B[34m"), END("\u001B[39m");
```
Comment (pkliczewski) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java:7]: It seems to me that we are mixing few things here. We have log levels (info, warn, err) and not log levels text and end. I think we should split

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/entity/WorkFlowTaskExecutionLog.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+public class WorkFlowTaskExecutionLog extends AbstractEntity {
+
+	@OneToOne(optional = false, fetch = FetchType.LAZY)
+	@JoinColumn(name = "workflow_task_execution_id")
+	private WorkFlowTaskExecution workFlowTaskExecution;
+
+	private String log;
+
+	public void addLog(String nextLog) {
+		log = log + "\n" + nextLog;
```
Comment (pkliczewski) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/entity/WorkFlowTaskExecutionLog.java:52]: Why do we want to keep logs in single line?

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowTaskLog.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+@NoArgsConstructor
+public class WorkFlowTaskLog {
+
+	private String logText;
+
+	private LogType logType = LogType.TEXT;
+
+	@Override
+	public String toString() {
+		return logType.getCode() + logText + (logType.getCode().isEmpty() ? "" : LogType.END.getCode());
```
Comment (pkliczewski) [inline parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowTaskLog.java:21]: do we care about timestamp? How do we know what was the order of logs?

**Comment by eloycoto** [comment]
I do not really like the implementation TBH. As a user experience, I think that it's not great. For me, the implementation should be something. 


this.log.error()
this.log.info()

And execute context should be inherent inside the class, and the log should log the workflow.ID as wid='context.get(workflowid)' In this way, a user can retrieve all the logs related to a workflow based on a simple query in any 3-party tool (Loki, ElasticSearch, etc..)

Dancing the context with multiple functions is not the best, and a function like `addTaskLog` is not great for user experience. 


But the code lgtm.

**Comment by RichardW98** [comment]
> 

agree.  from user experience, we should simplify the code to get context property.  I will fix it in next PR.

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/entity/WorkFlowTaskExecutionLog.java (hunk 2 above)
Comment (RichardW98) [reply workflow-service/src/main/java/com/redhat/parodos/workflow/execution/entity/WorkFlowTaskExecutionLog.java:52]: I will remove \n

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java (hunk 1 above)
Comment (RichardW98) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java:7]: > I do not really like the implementation TBH. As a user experience, I think that it's not great. For me, the implementation should be something.
> 
> this.log.error() this.log.info()
> 
> And execute context should be inherent inside the class, and the log should log the workflow.ID as wid='context.get(workflowid)' In this way, a user can retrieve all the logs related to a workflow based on a simple query in any 3-party tool (Loki, ElasticSearch, etc..)
> 
> Dancing the context with multiple functions is not the best, and a function like `addTaskLog` is not great for user experience.
> 
> But the code lgtm.

https://issues.redhat.com/browse/FLPATH-403

**Comment by RichardW98** [comment]
> 

https://issues.redhat.com/browse/FLPATH-403

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowLogServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			log.error("failed updating task execution for: {} in execution: {}. error Message: {}",
+					workFlowTaskExecution.getWorkFlowTaskDefinitionId(), workFlowTaskExecution.getId(), e.getMessage());
+			throw new WorkflowPersistenceFailedException(e.getMessage());
+		}
+	}
+
+	private WorkFlowTaskExecution getWorkFlowTaskExecution(UUID mainWorkflowExecutionId, String taskName) {
+		WorkFlowTaskDefinition workFlowTaskDefinition = Optional
+				.ofNullable(workFlowTaskDefinitionRepository.findFirstByName(taskName))
+				.orElseThrow(() -> new ResourceNotFoundException("task is not found!"));
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowLogServiceImpl.java:68]: Pls extend ResourceType and use 
```
ResourceNotFoundException(ResourceType.WORKFLOW_TASK, IDType.NAME, taskName)
```

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowLogServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		WorkFlowExecution workFlowExecution = Optional
+				.ofNullable(workFlowRepository.findFirstByMainWorkFlowExecutionIdAndWorkFlowDefinitionId(
+						mainWorkflowExecutionId, workFlowTaskDefinition.getWorkFlowDefinition().getId()))
+				.orElseThrow(() -> new ResourceNotFoundException("workflow execution is not found!"));
+		return Optional
+				.ofNullable(workFlowTaskRepository.findByWorkFlowExecutionIdAndWorkFlowTaskDefinitionId(
+						workFlowExecution.getId(), workFlowTaskDefinition.getId()))
+				.filter(workFlowTaskExecutions -> !workFlowTaskExecutions.isEmpty())
+				.map(workFlowTaskExecutions -> workFlowTaskExecutions.get(0))
+				.orElseThrow(() -> new ResourceNotFoundException("task execution is not found!"));
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/service/WorkFlowLogServiceImpl.java:78]: and here.

File: workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java [hunk 6]
```diff
@@ -148,4 +153,16 @@ public ResponseEntity<WorkFlowContextResponseDTO> getWorkflowParameters(@PathVar
...
 
+	@Operation(summary = "Returns workflow execution log")
+	@ApiResponses(value = {
+			@ApiResponse(responseCode = "200", description = "Succeeded",
+					content = { @Content(mediaType = MediaType.TEXT_PLAIN_VALUE,
+							schema = @Schema(implementation = String.class)) }),
+			@ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content),
+			@ApiResponse(responseCode = "403", description = "Forbidden", content = @Content) })
+	@GetMapping("/{workFlowExecutionId}/{taskName}/log")
+	public ResponseEntity<String> getLog(@PathVariable UUID workFlowExecutionId, @PathVariable String taskName) {
```
Comment (masayag) [inline workflow-service/src/main/java/com/redhat/parodos/workflow/execution/controller/WorkFlowController.java:164]: pls add tests to WorkFlowControllerTest

**Review by masayag — COMMENTED** [review]
Generally speaking,  would we like to consider alternative storage to store the data in the DB?

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RichardW98) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java:7]: to set log level and its ansi code for color

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.redhat.parodos.workflow.enums;
+
+import lombok.Getter;
+
+public enum LogType {
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/enums/LogType.java:5]: I will rename to loglevel

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowLogService.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.redhat.parodos.workflow.log;
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowLogService.java:1]: ok I will move to log.service and log.dto

File: parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowLogService.java [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+import java.util.UUID;
+
+public interface WorkFlowLogService {
+
+	String getLog(UUID workflowExecutionId, String taskName);
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply parodos-model-api/src/main/java/com/redhat/parodos/workflow/log/WorkFlowLogService.java:7]: yep, naming convention is `workFlow***`

File: workflow-service/generated/openapi/openapi.json [hunk 10]
```diff
@@ -929,6 +929,77 @@
         "summary" : "Returns a workflow status",
         "tags" : [ "Workflow" ]
       }
+    },
+    "/api/v1/workflows/{workFlowExecutionId}/{taskName}/log" : {
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply workflow-service/generated/openapi/openapi.json:933]: currently the log is only shown for each task node. But I can change it for potential future usage

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Review by anludke — APPROVED** [review]

**Pull request merged**